### PR TITLE
User session management

### PR DESCRIPTION
**Pull request by jonathan-952**
added session initialization and authentication middleware that will:
- check if a user is logged in or not
- check if a session is expired
- create session that connects user id to session

image
image

### CODE REVIEW

File: backend/.env
Comment (fahadakhan2) [inline backend/.env:1]: We should not be pushing our .env to github that is a big security risk. Please remove it from github.

File: backend/prisma/schema.prisma [hunk 1]
```diff
@@ -22,8 +22,8 @@ model User {
...
 model Session {
-  id        String   @id
-  sid       String   @unique
-  data      String
-  expiresAt DateTime
-} 
\ No newline at end of file
+  id          String      @id
+  sid         String      @unique
+  data        String?
```
Comment (fahadakhan2) [inline backend/prisma/schema.prisma]: I don't think the data should be optional so remove the question mark

File: backend/routes/userSession.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+const userSessionRouter = express.Router();
+
+const userSession = (session({
+    secret: process.env.SESSION_SECRET,
+    saveUninitialized: false,
+    resave: false,
+    cookie: {
+        secure: false,
+        httpOnly: true,
+        maxAge: 1000 * 60 * 30
```
Comment (fahadakhan2) [inline backend/routes/userSession.js]: nit: extract the 1000 * 60 * 30 into a variable that explains the time you are trying to achieve such as THREE_DAYS. Better for readability.

File: backend/routes/userSession.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}));
+
+const authenticateSession = ( async (req, res, next) => {
+    if (!req.session) {
+        return res.status(401).json({message: "not logged in"})
+    }
+    if (req.session.cookie.expires && req.session.cookie.expires < Date.now()) {
+        req.session.destroy((err) => {
+            if (err) {
+                console.error('error updating session')
```
Comment (fahadakhan2) [inline backend/routes/userSession.js]: probably want to remove this code since its only erroing to console?

File: backend/routes/userSession.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        prisma,
+        {
+            checkPeriod: 2 * 60 * 1000,  //ms
+            dbRecordIdIsSessionId: true,
+            dbRecordIdFunction: undefined,
+        },
+    )
+}));
+
+const authenticateSession = ( async (req, res, next) => {
```
Comment (fahadakhan2) [inline backend/routes/userSession.js:28]: this function can be cleaned up there is a ton of nested trys and conditional statements. See if you can simplify this logic and make it more readable

File: backend/routes/usernameRoute.js [hunk 5]
```diff
@@ -19,7 +20,7 @@ const checkUsernameExists = async (req, res, next) => {
     }
 }
 
-usernameRouter.put('/update-user', checkUsernameExists, async (req, res) => {
+usernameRouter.put('/update-user', authenticateSession, checkUsernameExists, async (req, res) => {
```
Comment (fahadakhan2) [inline backend/routes/usernameRoute.js]: the route handler should be extracted into a different file or function and you can just do what you are doing with the middleware and call it by its name.

File: backend/server.js [hunk 6]
```diff
@@ -1,39 +1,78 @@
...
-            dbRecordIdFunction: undefined,
+app.use(cors({
+    origin: 'http://localhost:3000/',
+    credentials: true
+  }));
+app.use('/reset-username-form', usernameRouter)
+app.use('/user-session', userSessionRouter)
+app.use(userSession)
+
+app.post('/sign-in', async (request, response) => { // Sign in route, get the users data to login
```
Comment (fahadakhan2) [inline backend/server.js]: extract the route handler function elsewhere so you can call it by name and try simplifying the logic. There is a ton of try catches and conditionals and there is probably some logic that is duplicate/ not useful which you can remove.

File: backend/.gitignore [hunk 7]
```diff
@@ -1,3 +1,3 @@
 /node_modules
 .env
-/database/prisma/dev.db
\ No newline at end of file
+/database/prisma/dev.dbgit
```
Comment (chinkevin25) [inline backend/.gitignore]: this got changed on accident it seems?

File: backend/.env [hunk 8]
```diff
@@ -1 +1,2 @@
-DATABASE_URL = "file:./dev.db"
\ No newline at end of file
+DATABASE_URL = "file:./dev.db"
+SESSION_SECRET = '6qyI+87A_?=;dw'
```
Comment (chinkevin25) [inline backend/.env:2]: this is still showing up in the diff. you might need to run something like https://www.wisdomgeek.com/development/how-to-permanently-remove-a-file-from-git-history/

File: backend/handlers/createSession.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+
+const createSession = ( async (request, response) => {
```
Comment (chinkevin25) [inline backend/handlers/createSession.js]: are we not using express-session? We shouldn't have to manually implement session creation logic if we're using that package.

File: backend/routes/userSession.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    store: new PrismaSessionStore(
+        prisma,
+        {
+            checkPeriod: 2 * 60 * 1000,  //ms
+            dbRecordIdIsSessionId: true,
+            dbRecordIdFunction: undefined,
+        },
+    )
+}));
```
Comment (chinkevin25) [inline backend/routes/userSession.js]: this should be creating your sessions for you already. What are you trying to use createSession for?

File: backend/routes/userSession.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    if (req.session.cookie.expires && req.session.cookie.expires < Date.now()) {
+        req.session.destroy()
+        return res.status(401).json({message: 'session expired'})
+    }
+
+    const userID = req.session.user;
+    try {
+        const session = await prisma.session.findUnique({
+            where: {id: userID}
+        });
```
Comment (chinkevin25) [inline backend/routes/userSession.js]: we shouldn't be doing it like this! we should be getting a session ID from the cookie right, we need to check that the cookie's session id exist, not find a session that is tied to a user.

conceptually you'll use `sid` to find session, then check that the session data's user exists and attach it to the request. 

In actuality, express-session should take care of it finding the session from the database already and have it attached as `req.session` so you shouldn't actually have to find your session manually.

i.e. this find session block is not needed iirc

File: backend/server.js [hunk 12]
```diff
@@ -2,38 +2,32 @@ if (process.env.NODE_ENV !== 'production') { // can load .env file while server
...
-const {PrismaClient} = require('@prisma/client')
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+const port = 8080;
+const {authenticateSession, userSession, userSessionRouter} = require('./routes/userSession.js');
 const usernameRouter = require('./routes/usernameRoute');
-app.use(cors({ origin:'http://localhost:3000'}));
-app.use('/reset-username-form', usernameRouter)
+const authHelper = require('./authHelper')
+import {createSession} from './handlers/createSession.js';
```
Comment (chinkevin25) [inline backend/server.js]: can you help me understand what this is doing?

Comment (chinkevin25) [reply backend/server.js]: actually doesn't this line throw an error if you run the server? import is not a recognized keyword in the backend without additional set up.

File: backend/handlers/createSession.js (hunk 9 above)
Comment (chinkevin25) [reply backend/handlers/createSession.js]: ah this is not creating a session -- it's loggin in right? we can keep we should just name it more clearly. I thought you were trying to create a session.

File: backend/handlers/createSession.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+            return response.status(404).json({message: "user not found"}) 
+        }
+        const passwordMatch = await bcrypt.compare(password, user.password)
+    
+        if (passwordMatch) {
+            request.session.user = user.id
+        }
+        else {
+            return response.status(400).send('invalid email or password')
+        }
```
Comment (chinkevin25) [inline backend/handlers/createSession.js]: this is stylistic but typically we want to improve readability by reducing nesting.

```
if (!passwordMatch) {
  return response.status(400).send('invalid email or password')
}

request.session.user = user.id
```

reduces the nesting by one. it's a common pattern 

no change needed though

File: frontend/src/handlers/handleResetUsername.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        setMessage('fill in all required information')
+        return;
+    }
+    const usernameExists = await checkUsername();
+    if (usernameExists) {
+        setMessage("username already exists")
+        return;
+    }
+    try {
+        const res = await axios.put(`http://localhost:8080/reset-username-form/update-user/`, {
```
Comment (chinkevin25) [inline frontend/src/handlers/handleResetUsername.js:17]: nit. i think you have proxy set up already so you don't need to specify the localhost:8080

File: backend/routes/userSession.js (hunk 11 above)
Comment (jonathan-952) [reply backend/routes/userSession.js]: i worked out the stuff with checking for a cookie, im confused about where you said, "then check that the session data's user exists and attach it to the request.". i am attaching the user id to the request on the login, am i doing it wrong then or is it a misunderstanding

File: backend/server.js (hunk 12 above)
Comment (jonathan-952) [reply backend/server.js]: im not sure what you mean. are you talking about the lines in the green or red?

File: backend/.gitignore (hunk 7 above)
Comment (jonathan-952) [reply backend/.gitignore]: i dont think this was me unless a misclick. dont remember changing this though

File: backend/handlers/createSession.js (hunk 9 above)
Comment (jonathan-952) [reply backend/handlers/createSession.js]: my bad for the bad naming. i renamed it to login handler. whoops

File: backend/.env (hunk 8 above)
Comment (jonathan-952) [reply backend/.env:2]: followed all the steps. not sure how to check if it worked though

**Pull request merged**